File: LeetCode/Python3/Problem 131.py
# LeetCode: Problem 131
# Jonathan Kosasih

#Given a string s, partition s such that every substring of the partition is a palindrome.
#
#Return all possible palindrome partitioning of s.
#
#Example:
#Input: "aab"
#Output:
#[
#  ["aa","b"],
#  ["a","a","b"]
#]
#------------------------------------------------------------------
class Solution(object):
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        memo = dict()
        return self.helper(s, memo)
    
    def helper(self, s, memo):
        if(s in memo):
            return memo[s]
        elif(not s):
            memo[s] = [[]]
        else:
            temp = []
            for i in range(1, len(s)+1):
                if(self.palindrome(s[:i])):
                    for partition in self.helper(s[i:], memo):
                        temp.append([s[:i]] + partition)
            memo[s] = temp
        return memo[s]
        
    def palindrome(self, s):
        return "".join(reversed(s)) == s
        
        
sol = Solution()
string = "aab"
print(str(sol.partition(string)))

File: LeetCode/C++/Problem 131.cpp
// LeetCode: Problem 131
// Jonathan Kosasih
/*
Given a string s, partition s such that every substring of the partition is a palindrome.

Return all possible palindrome partitioning of s.
*/
//-------------------------------------------------------------------

#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <set>
#include <string>
#include <algorithm> 
#include <stack>
#include <queue>
#include <deque>
using std::vector;
using std::map;
using std::unordered_map;
using std::set;
using std::stack;
using std::queue;
using std::deque;

class Solution {
public:
	vector<vector<string>> partition(string s) {
		vector<vector<string>> pars;
		vector<string> par;
		partition(s, 0, par, pars);
		return pars;
	}
private:
	void partition(string& s, int start, vector<string>& par, vector<vector<string>>& pars) {
		int n = s.length();
		if (start == n) {
			pars.push_back(par);
		}
		else {
			for (int i = start; i < n; i++) {
				if (isPalindrome(s, start, i)) {
					par.push_back(s.substr(start, i - start + 1));
					partition(s, i + 1, par, pars);
					par.pop_back();
				}
			}
		}
	}

	bool isPalindrome(string& s, int l, int r) {
		while (l < r) {
			if (s[l++] != s[r--]) {
				return false;
			}
		}
		return true;
	}
};
int main()
{
    //std::cout << "Hello World!\n";
}